File: googleDriveClone-kys/googleDrive/src/main/java/com/googledrive/config/SecurityConfig.java
package com.googledrive.config;

import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import java.util.List;

/**
 * Spring Security 설정.
 * - CSRF 비활성화 (REST API는 토큰 기반이므로 불필요)
 * - 세션 STATELESS (JWT 사용으로 서버에 세션 미저장)
 * - /api/auth/** 는 인증 없이 접근 허용
 * - 나머지는 인증 필요
 * - CORS: Vite 개발 서버(localhost:5173) 허용
 */
@Configuration
@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig {

    private final JwtAuthenticationFilter jwtAuthenticationFilter;

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        http
                // CSRF 비활성화
                .csrf(AbstractHttpConfigurer::disable)

                // CORS 설정 적용
                .cors(cors -> cors.configurationSource(corsConfigurationSource()))

                // 세션 미사용 (JWT 기반 인증)
                .sessionManagement(session ->
                        session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))

                // 요청별 접근 권한 설정
                .authorizeHttpRequests(auth -> auth
                        // 회원가입, 로그인, 비밀번호 찾기는 인증 없이 허용
                        .requestMatchers(
                                "/api/auth/signup",
                                "/api/auth/login",
                                "/api/auth/security-question",
                                "/api/auth/reset-password"
                        ).permitAll()
                        // 공유 링크 조회는 인증 없이 허용 (추후 파일 공유 기능용)
                        .requestMatchers(HttpMethod.GET, "/api/shares/**").permitAll()
                        // 나머지 모든 요청은 인증 필요
                        .anyRequest().authenticated()
                )

                // JWT 필터를 UsernamePasswordAuthenticationFilter 앞에 추가
                .addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class);

        return http.build();
    }

    /**
     * BCrypt 비밀번호 인코더 Bean.
     * AuthService에서 @Autowired로 주입받아 사용한다.
     */
    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    /**
     * CORS 설정.
     * - 허용 출처: http://localhost:5173 (Vite 개발 서버)
     * - 허용 메서드: GET, POST, PUT, PATCH, DELETE, OPTIONS
     * - 허용 헤더: 모든 헤더
     * - 자격증명(쿠키 등) 허용
     */
    @Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration config = new CorsConfiguration();
        config.setAllowedOrigins(List.of("http://localhost:5173"));
        config.setAllowedMethods(List.of("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"));
        config.setAllowedHeaders(List.of("*"));
        config.setAllowCredentials(true);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", config);
        return source;
    }
}

File: drive-jy/src/main/kotlin/demo/drive/common/config/SecurityConfig.kt
package demo.drive.common.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.access.hierarchicalroles.RoleHierarchy
import org.springframework.security.access.hierarchicalroles.RoleHierarchyImpl
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain

@Configuration
@EnableWebSecurity
class SecurityConfig {

    @Bean
    fun passwordEncoder(): PasswordEncoder = BCryptPasswordEncoder()

    /**
     * 역할 계층 정의. 새 역할 추가 시 여기만 수정.
     * ADMIN > USER — ADMIN은 USER의 모든 권한을 자동으로 포함한다.
     */
    @Bean
    fun roleHierarchy(): RoleHierarchy =
        RoleHierarchyImpl.fromHierarchy(
            """
            ROLE_ADMIN > ROLE_USER
            """.trimIndent()
        )

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .authorizeHttpRequests {
                it.requestMatchers("/auth/**", "/share/**", "/api/share/resolve/**", "/css/**", "/js/**", "/icons/**").permitAll()
                it.requestMatchers("/h2-console/**").permitAll()
                it.anyRequest().authenticated()
            }
            .formLogin {
                it.loginPage("/auth/login")
                it.loginProcessingUrl("/auth/login")
                it.usernameParameter("email")
                it.passwordParameter("password")
                it.defaultSuccessUrl("/drive", true)
                it.failureUrl("/auth/login?error")
            }
            .logout {
                it.logoutUrl("/auth/logout")
                it.logoutSuccessUrl("/auth/login")
            }
            .rememberMe {
                it.key("drive-remember-me")
                it.tokenValiditySeconds(604800)
            }
            .csrf { it.disable() }
            .headers {
                it.frameOptions { fo -> fo.sameOrigin() }
            }
        return http.build()
    }
}
